bellatrix.core/src/main/java/solutions/bellatrix/core/utilities/HtmlService.java

```diff
@@ -0,0 +1,167 @@
+    }
+
+    public static String findRelativeCssLocator(Element element, String xpath) {
+        var absoluteXpath = HtmlService.findRelativeElementAbsoluteXpath(element, xpath);
+
+        return HtmlService.convertXpathToCss(absoluteXpath);
+    }
+
+    public static ArrayList<String> findCssLocators(String html, String xpath) {
+        var absoluteXpaths = HtmlService.findElementsAbsoluteXpath(html, xpath);
+
+        return HtmlService.convertXpathToCss(absoluteXpaths);
+    }
+
+    public static ArrayList<String> findRelativeCssLocators(Element element, String xpath) {
+        var absoluteXpaths = HtmlService.findRelativeElementsAbsoluteXpath(element, xpath);
+
+        return HtmlService.convertXpathToCss(absoluteXpaths);
+    }
+
+    public static Element findElement(String html, String cssQuery) {
+        var doc = Jsoup.parse(html);
+
+        return doc.select(cssQuery).stream().findFirst()
+                .orElseThrow(() -> new IllegalArgumentException(String.format("No element was found with the css: %s", cssQuery)));
+    }
+
+    private static String findElementAbsoluteXpath(String html, String xpath) {
+        var doc = Jsoup.parse(html);
+        var el = doc.selectXpath(xpath);
+
+        if (el.isEmpty()) {
+            throw new IllegalArgumentException(String.format("No element was found with the xpath: %s", xpath));
+        }
+
+        return getAbsoluteXPath(el.first());
+    }
+
+    private static ArrayList<String> findElementsAbsoluteXpath(String html, String xpath) {
+        var doc = Jsoup.parse(html);
+        var elements = doc.selectXpath(xpath).stream().toList();
+
+        if (elements.isEmpty()) {
+            throw new IllegalArgumentException(String.format("No elements were found with the xpath: %s", xpath));
+        }
+
+        ArrayList<String> individualQueries = new ArrayList<>();
+
+        for (Element el : elements) {
+            individualQueries.add(getAbsoluteXPath(el));
+        }
+
+        return individualQueries;
+    }
+
+    private static String findRelativeElementAbsoluteXpath(Element baseElement, String xpath) {
+        var el = baseElement.selectXpath(xpath);
+
+        if (el.isEmpty()) {
+            throw new IllegalArgumentException(String.format("No element was found with the xpath: %s", xpath));
+        }
+
+        return getAbsoluteXPath(el.first());
+    }
+
+    private static ArrayList<String> findRelativeElementsAbsoluteXpath(Element baseElement, String xpath) {
+        var elements = baseElement.selectXpath(xpath).stream().toList();
+
+        if (elements.isEmpty()) {
+            throw new IllegalArgumentException(String.format("No elements were found with the xpath: %s", xpath));
+        }
+
+        ArrayList<String> queries = new ArrayList<String>();
+
+        for (Element el : elements) {
+            queries.add(getAbsoluteXPath(el));
+        }
+
+        return queries;
+    }
+
+    private static String convertXpathToCss(String xpath) {
```

Comment:
this method is the core of this functionality, but it does not seem to cover many of the cases at first glance:

- contains(text(),'sampleText')
- //node[attribute='value']//descendant::attribute
- /last()
- ends-with()
- and so on

Reply:
This method `convertXpathToCss(String xpath)` works with already processed xpath;
The logic of this service is as follows:
1. Use `JSoup` to find the `Element` inside the html document (the html that is being passed is `element.shadowRoot.innerHTML`, the xpath being our original xpath)
2. After we have found the Element, we get it's absolute position inside the html document (absolute xpath)
3. using `convertXpathToCss(String xpath)`, we convert this absolute xpath to absolute css
4. we return it to the module (`web/playwright`)

Although it may sound slightly counter-intuitive to use absolute xpath/absolute css, they are always dynamically generated from the actual xpath;

The limitations of the xpaths we can use are set by `JSoup` capability of parsing them
Other than that, every function and axis is possible

Reply:
We could work on refactoring this service to be more understandable.